### PR TITLE
BLD/RLS: Update wheels to include GDAL 3.9.2 and libkml

### LINKED ISSUE #378: To include spatialite in wheels or not to?
As mentioned here, https://github.com/geopandas/geopandas/pull/3223, think through wheter it it would be better to include it or not...

### LINKED ISSUE #447: RLS: add LIBKML to GDAL vcpkg config for release
Per #444, we should probably include `LIBKML` in our VCPKG config, it is available as an [option](https://vcpkg.link/ports/gdal).

### LINKED ISSUE #453: BLD/RLS: manylinux2014 at end of life
Per [manylinux EOL table](https://github.com/mayeut/pep600_compliance/blob/master/EOL.rst), manylinux2014 is at end of life.  We are currently using this for building x86_64 wheels, but using manylinux_2_28 for Arm64 wheels.

It appears the end of life date is set based on the Centos 7 EOL date.  The Docker image for manylinux2014 has been updated recently to resolve issues with Centos reaching EOL (fetching packages from Centos mirrors was failing), so it may continue to work for some time.

We have to use a custom install of zlib via VCPKG in manylinux2014 to get around compatibility issues (see #57), but don't have to do those same workarounds in manylinux_2_28.  This presents further issues as we try to add support for LIBKML in #589, which requires minizip, which depends on zlib - and probably requires a custom port of minizip for that to work.

It looks like manylinux_2_28 is the [recommended distro](https://github.com/mayeut/pep600_compliance?tab=readme-ov-file#acceptable-distros-to-build-wheels) for building wheels now, so we should plan to switch to this unless there is a reason not to.

### LINKED ISSUE #454: BUG: wheel tests on macos-latest use the latest Arm64 version of pyogrio from PyPI
See logs from 0.9.0 release action [here](https://github.com/geopandas/pyogrio/actions/runs/9553600977/job/26334086179).  

The `macos-latest` runner is now MacOS 14 / Arm64 only, whereas previously it was x86_64.

Based on how release tests are setup, we are manually downloading the x86_64 wheel from the action artifacts to a `macos-latest` runner, but then installing that wheel fails because of architecture mismatch, so pip falls back to installing the latest Arm64 wheel from PyPI - which obviates the test altogether.

Instead, we need to have the `macos-latest` runner download the Arm64 wheel from the release artifacts instead and test with that.  We also should try to make sure that pip doesn't automatically & silently fall back to installing pyogrio from PyPI.

### PR DESCRIPTION
**Pull request by brendan-ward**
(edit: changed 3.9.1 to 3.9.2)

resolves #447, #453, #454

Uses GDAL 3.9.2 from VCPKG, which requires bumping this to the next minor release (0.10.0), and enables the libkml features (resolves #447).  Also adds tests for GDAL 3.9.2 to CI.

This drops use of manylinux2014 and uses manylinux_2_28 (minimum recommended version) for x86_64 builds (resolves #453); we were already using it for Arm64 builds.  This drops the custom port of zlib, which no longer appears necessary (was due to version conflicts with manylinux2014) and was causing issues adding one of the dependencies of libkml.  Switching to manylinux_2_28 also obviates the build error we've been seeing lately due to CentOS 7 (used by manylinux2014) EOL.

This switches to pulling the Arm64 MacOS wheel when testing on the `macos-latest` runner, which is MacOS / Arm64 (resolves #454) and uses `--no-index` flag to `uv pip install` to ensure it doesn't silently fall back to using an Arm64 wheel from PyPI.

This switches to `uv` for installing Python dependencies, in line with #438, which also gets around issues with Python 3.12 requiring a virtual environment (needed for testing the sdist using the `gdal:ubuntu-small-3.9.2` base image which has Python 3.12).

I initially tried to add libspatialite, but was unsuccessful and getting that to build within VCPKG.

### CODE REVIEW

**Comment by brendan-ward** [comment]
Adding LIBKML appears to require a number of boost libraries, which may make the wheels a bit bigger:

```bash
  * boost-assert:arm64-linux@1.85.0#1
  * boost-cmake:arm64-linux@1.85.0#1
  * boost-config:arm64-linux@1.85.0#1
  * boost-core:arm64-linux@1.85.0#2
  * boost-headers:arm64-linux@1.85.0#1
  * boost-move:arm64-linux@1.85.0#1
  * boost-smart-ptr:arm64-linux@1.85.0#1
  * boost-static-assert:arm64-linux@1.85.0#1
  * boost-throw-exception:arm64-linux@1.85.0#1
  * boost-type-traits:arm64-linux@1.85.0#1
  * boost-uninstall:arm64-linux@1.85.0#1
    libkml:arm64-linux@1.3.0#13
  * minizip:arm64-linux@1.3.1#1
  * uriparser:arm64-linux@0.9.8
  * vcpkg-boost:arm64-linux@2024-05-15
```

spatialite is creating some dependency issues: it brings in freexl by default, which is failing to build.  I don't know that we actually need it to support spatialite functionality in GDAL / pyogrio.  (need to disable default features for spatialite)

**Comment by brendan-ward** [comment]
I can't seem to get libspatialite to build within vcpkg locally, even if I just try to build it directly and disable default features.  It appears to be failing to detect sqlite3, even though that gets installed properly by vcpkg.  Unless anyone has some brilliant ideas how to get this working, I think we should leave libspatialite out for now.

**Comment by brendan-ward** [comment]
We appear to have jobs stuck in the github actions queue for a long time (due to github actions outage / issues yesterday), and canceling them does not clear them out.  I'm going to temporarily close this PR to see if that triggers github to clear out those jobs.

**Comment by EwoutH** [comment]
Since it looks like this PR is needed for Python 3.13 wheels, what's needed to move it forward?

**Comment by EwoutH** [comment]
GDAL 3.9.2 was [released](https://github.com/OSGeo/gdal/releases/tag/v3.9.2) with additional bug fixes.

**Comment by brendan-ward** [comment]
Failures using Conda appear related to [shapely #2098](https://github.com/shapely/shapely/issues/2098), which will be resolved when shapely 2.0.6 lands on conda-forge.

**Comment by EwoutH** [comment]
A nice badge to view when Shapely 2.0.6 is available on conda-forge:

[Anaconda](https://anaconda.org/conda-forge/shapely)

**Comment by theroggy** [comment]
The following PR will avoid the conda environment to be cached (for a very long time). This will lead to shapely 2.0.6 to be actually used in the tests: https://github.com/geopandas/pyogrio/pull/461.

**Comment by EwoutH** [comment]
Looks fully green!

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -54,26 +54,34 @@ jobs:
...
+        run: |
+          apt-get update && apt-get install -y build-essential python3-dev
+
+      - name: Create virtual environment
+        run: |
+          curl -LsSf https://astral.sh/uv/install.sh | sh
+          . $HOME/.cargo/env
+          uv venv .venv
+          echo "VIRTUAL_ENV=.venv" >> $GITHUB_ENV
+          echo "$PWD/.venv/bin" >> $GITHUB_PATH
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply .github/workflows/release.yml:70]: But I suppose it is for the `python3 -m pytest ..` below? That could maybe also use `uv run python -m pytest ...` to make it explicit it is using the python from the uv venv?

File: CHANGES.md [hunk 2]
```diff
@@ -18,6 +18,13 @@
 -   Raise NotImplmentedError when user attempts to write to an open file handle (#442).
 -   Prevent seek on read from compressed inputs (#443).
 
+### Packaging
```
Comment (jorisvandenbossche) [inline CHANGES.md:21]: There is already a packaging section above, so you can merge that

File: ci/manylinux2014_x86_64-vcpkg-gdal.Dockerfile
Comment (jorisvandenbossche) [inline ci/manylinux2014_x86_64-vcpkg-gdal.Dockerfile:1]: If this file is no longer used, it can be removed?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -178,7 +186,7 @@ jobs:
           git reset --hard
           # pull latest version to ensure the required commit with GDAL 3.8 is available
           git pull
-          git checkout 4f4a1821b2e8c7a2863e4df65a4d514f84144049
+          git checkout 73794ce5f63fd138fab999a22959ca7c6305d93c
 
       - name: Install GDAL
         env:
```
Comment (jorisvandenbossche) [inline .github/workflows/release.yml:192]: There is a `--overlay-triplets=./ci/custom-triplets` a few lines below here that can then be removed as well

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for working on this!

File: .github/workflows/release.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (brendan-ward) [reply .github/workflows/release.yml:70]: Correct, I believe I was running into issues with `uv` not automatically activating the virtual environment in subsequent steps.  Somehow I missed that you could issue `uv run python ...` instead; will try that now.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -54,26 +54,34 @@ jobs:
     needs: [build-sdist]
     runs-on: ubuntu-latest
     container:
-      image: "ghcr.io/osgeo/gdal:ubuntu-small-3.8.5"
+      image: "ghcr.io/osgeo/gdal:ubuntu-small-3.9.2"
 
     steps:
+      - name: Install packages
+        run: |
+          apt-get update && apt-get install -y build-essential python3-dev
```
[the remark this answers is not in the document]
Comment (brendan-ward) [reply .github/workflows/release.yml:62]: Originally we were installing `python3-pip` which also pulled in `build-essential` and `python3-dev` as [dependencies](https://packages.ubuntu.com/jammy/python3-pip).  We now need to install them directly.

File: .github/workflows/release.yml (hunk 3 above)
Comment (brendan-ward) [reply .github/workflows/release.yml:192]: Don't we still need to use the custom triplets?

File: .github/workflows/release.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (brendan-ward) [reply .github/workflows/release.yml]: Just dropping a note that trying `uv run python` did not work in this case (appears to require `[project]` section be added to `pyproject.toml` plus a few more things I didn't investigate fixing); it was easier to just use the environment variables to activate the virtual environment.

**Comment by brendan-ward** [comment]
Superseded by #466

**Pull request closed without merging**